src/spss/readstat_sav_read.c

```diff
@@ -145,6 +147,212 @@ static readstat_error_t sav_parse_variable_display_parameter_record(sav_ctx_t *c
+        goto cleanup;
+    }
+    next_part++; // Move past the space after the digits
+    if (strlen(next_part) < count) {
+        retval = READSTAT_ERROR_BAD_MR_STRING;
+        goto cleanup;
+    }
+
+    result->label = readstat_malloc(count + 1);  // +1 for the null-terminator
+    if (result->label == NULL) {
+        retval = READSTAT_ERROR_MALLOC;
+        goto cleanup;
+    }
+    strncpy(result->label, next_part, count);
+    result->label[count] = '\0';
+
+    next_part += count;
+    if (*next_part != ' ') {
+        retval = READSTAT_ERROR_BAD_MR_STRING;
+        goto cleanup;
+    }
+    next_part++;
+
+    char **subvariables = NULL;
+    int subvar_count = 0;
+    while (*next_part) {
+        if (*next_part == ' ') {  // Skip any extra spaces
+            next_part++;
+            continue;
+        }
+
+        const char *start = next_part;
+        while (*next_part && *next_part != ' ') {
+            next_part++;  // Move to the end of the current subvariable
+        }
+
+        size_t length = next_part - start;
+        char *subvariable = readstat_malloc(length + 1);  // Allocate memory for the subvariable
+        if (subvariable == NULL) {
+            retval = READSTAT_ERROR_MALLOC;
+            for (int i = 0; i < subvar_count; i++) {
+                free(subvariables[i]);
+                subvariables[i] = NULL;
+            }
+            free(subvariables);
+            goto cleanup;
+        }
+        strncpy(subvariable, start, length);
+        subvariable[length] = '\0';  // Null-terminate the string
+
+        char **temp = realloc(subvariables, (subvar_count + 1) * sizeof(char *));
+        if (temp == NULL) {
+            retval = READSTAT_ERROR_MALLOC;
+            free(subvariable);
+            for (int i = 0; i < subvar_count; i++) {
+                free(subvariables[i]);
+                subvariables[i] = NULL;
+            }
+            free(subvariables);
+            goto cleanup;
+        }
+        subvariables = temp;
+        subvariables[subvar_count++] = subvariable;  // Add the new subvariable to the array
+
+        if (*next_part == ' ') next_part++; // Move past the space
+    }
+
+    result->subvariables = subvariables;
+    result->num_subvars = subvar_count;
+
+cleanup:
+    return retval;
+}
+
+static readstat_error_t sav_read_multiple_response_sets(size_t data_len, sav_ctx_t *ctx) {
+    readstat_error_t retval = READSTAT_OK;
+
+    char *mr_string = readstat_malloc(data_len + 1);
+    if (mr_string == NULL) {
+        retval = READSTAT_ERROR_MALLOC;
+        goto cleanup;
+    }
+    mr_string[data_len] = '\0';
+    if (ctx->io->read(mr_string, data_len, ctx->io->io_ctx) < data_len) {
+        retval = READSTAT_ERROR_PARSE;
+        goto cleanup;
+    }
+    if (mr_string[0] != '$') {
+        retval = READSTAT_ERROR_BAD_MR_STRING;
+        goto cleanup;
+    }
+
+    fprintf(stderr, "\n\n\nDebug: MR string: '%s'\n", mr_string);
+    char *token = strtok(mr_string, "$\n");
```

Comment:
I believe `strtok` is not thread-safe; I'd prefer a thread-safe implementation.

Reply:
Do you have any particular advice for this? I tried with `strtok_r` and `strtok_s` combination, but build system is affected... I guess one option would be to do the entire mr_string parser in ragel... But maybe there's a quicker way to do it.

Reply:
What was the build error?